File: Jongryul/01/12852.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim();
const n = Number(input);

const dp = Array.from({ length: n + 1 }, (_, i) => n - i);
const parent = Array.from({ length: n + 1 }, (_, i) => i);
const rlt = [1];
let idx = 1;

for (let i = n; i > 0; i--) {
  if (i != n && dp[i] >= dp[i + 1] + 1) {
    parent[i] = i + 1;
    dp[i] = dp[i + 1] + 1;
  }
  if (i / 3 === Math.ceil(i / 3) && dp[i / 3] > dp[i] + 1) {
    parent[i / 3] = i;
    dp[i / 3] = dp[i] + 1;
  }
  if (i / 2 === Math.ceil(i / 2) && dp[i / 2] > dp[i] + 1) {
    parent[i / 2] = i;
    dp[i / 2] = dp[i] + 1;
  }
}

while (idx !== n) {
  rlt.push(parent[idx]);
  idx = parent[idx];
}
rlt.reverse();
console.log(dp[1]);
console.log(rlt.join(" "));

File: 01_Season/10th_Competition/Jongryul/01/12852.py
n = int(input())

dp = [n - i for i in range(n+1)]
parents = [i for i in range(n+1)]

for i in range(n, 0, -1):

    if i < n and dp[i] >= dp[i+1] + 1:
        parents[i] = i+1
        dp[i] = dp[i+1] + 1

    if i / 3 == i // 3 and dp[i//3] > dp[i] + 1:
        parents[i//3] = i
        dp[i//3] = dp[i] + 1

    if i / 2 == i // 2 and dp[i//2] > dp[i] + 1:
        parents[i//2] = i
        dp[i//2] = dp[i] + 1


print(dp[1])
rlt = '1'
idx = parents[1]

while True:
    rlt = f'{idx} {rlt}'
    if idx == n:
        break
    idx = parents[idx]
if n != 1:
    print(rlt)
else:
    print(1)
